<!DOCTYPE html>
<html lang="en">

<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
</head>

<body>
  <script>
    /**
     * promise.then()
     * 返回一个Promise对象
     * 其状态由回调函数决定
     * 1.若出现异常，则为rejected，结果为reason
     * 2.无返回值，则为resolved，结果为undefined
     * 3.返回值为非Promise对象，则为resolved，结果为返回值
     * 4.返回值为Promise对象，则then()的返回值就为该Promise
     * 5.没有回调函数，则返回调用者
     */
    let result = new Promise((resolve, reject) => {
      resolve('OK')
    }).then(value => {
      // throw 'Error.'
      // return 520
    }, reason => {

    }).catch(reason => { })
    console.log(result)

    // 异常穿透
    new Promise((resolve, reject) => {
      reject('Error.')
    }).then(value => {
      console.log(111)
    }).then(value => {
      console.log(222)
    }).then(value => {
      console.log(333)
    }).catch(reason => {
      console.log(reason)
    })

    // 中断Promise链，返回一个pending状态的Promise
    new Promise((resolve, reject) => {
      resolve('Ok.')
    }).then(value => {
      console.log(111)
    }).then(value => {
      console.log(222)
      return new Promise(() => { })
    }).then(value => {
      console.log(333)
    })

  </script>
</body>

</html>
